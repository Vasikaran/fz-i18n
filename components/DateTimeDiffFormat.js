import React,{PropTypes, Children} from 'react';
import moment from "moment-timezone";

export default class DateTimeDiffFormat extends React.Component{
	constructor(props){
		super(props)
		this.getValues = this.getValues.bind(this);
		this.getSuffix = this.getSuffix.bind(this);
	}
	getSuffix(fromDateObj, toDateObj){
		let suffix;
		if(toDateObj.isBefore(fromDateObj)){
        	suffix=this.props.later || ''
        }else if(toDateObj.isAfter(fromDateObj)){
			suffix=this.props.ago || ''
        }else{
        	suffix=''
        }
        return suffix;
	}
	getValues(params=[],diff){
		return params.map((param)=>{
			return diff[param];
		})
	}
	render(){
		const { from, fromTimeZone, to, toTimeZone,today,yesterday,tomorrow,others } = this.props;
        let fromDateObj = moment(from).tz(fromTimeZone);
        let toDateObj = moment(to).tz(toTimeZone);
        let suffix = this.getSuffix(fromDateObj,toDateObj);
     	let diff = moment.duration(Math.abs(toDateObj.diff(fromDateObj)));
     	var diffObj = {
     		hh:pad(diff.get('h'),2),
     		mm:pad(diff.get('m'),2),
     		ss:pad(diff.get('s'),2),
     		MM:pad(diff.get('M'),2),
     		yy:pad(diff.get('y'),2),
     		dd:pad(diff.get('d'),2)
     	}
     	var daysDiff = toDateObj.diff(fromDateObj, 'days');
     	var dateTimeFormat;
     	if(daysDiff < -1){
     		dateTimeFormat = <span>less than one day</span>
     	}else if(daysDiff < 0){
     		dateTimeFormat = 
	     		<FormatText i18NKey={yesterday.key} values={this.getValues(yesterday.params,diffObj)}/>
     	}else if(daysDiff < 1){
     		dateTimeFormat = <span>
	     		<FormatText i18NKey={today.key} values={this.getValues(today.params,diffObj)}/>
	     		{' '}<FormatText i18NKey={suffix}/>
     		</span>
     	}else if(daysDiff < 2){
     		dateTimeFormat = <FormatText i18NKey={tomorrow.key} values={this.getValues(tomorrow.params,diffObj)}/>
     	} else {

     	} 
       
		return <span>{dateTimeFormat} </span>
	}
}