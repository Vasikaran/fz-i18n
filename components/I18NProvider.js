import React,{PropTypes, Children} from 'react';

const emptyObj={};
export class I18NProvider extends React.Component{
	getChildContext() {
    return {i18n: this.props.i18n};
  }
  render(){
  	return Children.only(this.props.children)
  }

}

I18NProvider.defaultProps = {
	i18n:emptyObj
}
I18NProvider.childContextTypes = {
  i18n: PropTypes.object
};

export default I18N;