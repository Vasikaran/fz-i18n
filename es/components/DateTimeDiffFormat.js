import _typeof from 'babel-runtime/helpers/typeof';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Children } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import { pad, getSuffix, getValues, isToday, isYesterday, isTomorrow, isWithinAWeek } from '../utils';
import FormatText from './FormatText';

var DateTimeDiffFormat = function (_React$Component) {
  _inherits(DateTimeDiffFormat, _React$Component);

  function DateTimeDiffFormat(props) {
    _classCallCheck(this, DateTimeDiffFormat);

    var _this = _possibleConstructorReturn(this, (DateTimeDiffFormat.__proto__ || _Object$getPrototypeOf(DateTimeDiffFormat)).call(this, props));

    _this.getSuffix = _this.getSuffix.bind(_this);
    return _this;
  }

  _createClass(DateTimeDiffFormat, [{
    key: 'getSuffix',
    value: function getSuffix(fromDateObj, toDateObj) {
      var suffix = void 0;
      if (toDateObj.isBefore(fromDateObj)) {
        suffix = this.props.ago || '';
      } else if (toDateObj.isAfter(fromDateObj)) {
        suffix = this.props.later || '';
      } else {
        suffix = '';
      }
      return suffix;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          from = _props.from,
          fromTimeZone = _props.fromTimeZone,
          to = _props.to,
          toTimeZone = _props.toTimeZone,
          today = _props.today,
          yesterday = _props.yesterday,
          tomorrow = _props.tomorrow,
          others = _props.others,
          format = _props.format,
          _props$className = _props.className,
          className = _props$className === undefined ? null : _props$className,
          _props$title = _props.title,
          title = _props$title === undefined ? null : _props$title;

      var fromDateObj = moment(from).tz(fromTimeZone);
      var toDateObj = moment(to).tz(toTimeZone);
      var suffix = this.getSuffix(fromDateObj, toDateObj);
      var diff = moment.duration(Math.abs(toDateObj.diff(fromDateObj)));
      var yearRemaingdays;
      if (toDateObj.isBefore(fromDateObj)) {
        var tempTo = moment(to).tz(toTimeZone).add(diff.get('y'), 'years');
        yearRemaingdays = Math.abs(tempTo.diff(fromDateObj, 'days'));
      } else {
        var tempFrom = moment(fromDateObj).tz(fromTimeZone).add(diff.get('y'), 'years');
        yearRemaingdays = Math.abs(tempFrom.diff(toDateObj, 'days'));
      }
      var withInAWeak = isWithinAWeek(fromDateObj, toDateObj);

      var diffObj = {
        h: diff.get('h'),
        m: diff.get('m'),
        s: diff.get('s'),
        M: diff.get('M'),
        y: diff.get('y'),
        d: diff.get('d'),
        hh: pad(diff.get('h'), 2),
        mm: pad(diff.get('m'), 2),
        ss: pad(diff.get('s'), 2),
        MM: pad(diff.get('M'), 2),
        yy: pad(diff.get('y'), 2),
        dd: pad(diff.get('d'), 2),
        days: yearRemaingdays,
        yDays: pad(yearRemaingdays, 2),
        isWithInAWeek: withInAWeak,
        suffix: suffix
      };
      var diffObj1 = {
        hours: diff.get('h'),
        minutes: diff.get('m'),
        seconds: diff.get('s'),
        months: diff.get('M'),
        years: diff.get('y'),
        days: diff.get('d'),
        yDays: yearRemaingdays,
        isWithInAWeek: withInAWeak,
        suffix: suffix
      };
      //var daysDiff = toDateObj.diff(fromDateObj, 'days', true);
      var key = '';
      var values = [];
      var text = null;
      var isSuffixEnable = false;
      if (format) {
        var years = void 0,
            months = void 0,
            days = void 0,
            hours = void 0,
            minutes = void 0,
            seconds = void 0;
        years = diffObj1.years > 1 ? '2' : diffObj1.years;
        //  months = diffObj1.months > 1 ? '2' : diffObj1.months;
        //days = diffObj1.days > 1 ? '2' : diffObj1.days;
        days = yearRemaingdays > 1 ? '2' : yearRemaingdays;
        hours = diffObj1.hours > 1 ? '2' : diffObj1.hours;
        minutes = diffObj1.minutes > 1 ? '2' : diffObj1.minutes;
        //seconds = diffObj1.seconds > 1 ? '2' : diffObj1.seconds;
        // let pattern = '' + years + months + days + hours + minutes + seconds;
        var count = 0;
        var pattern = [years, days, hours, minutes].reduce(function (res, next) {
          if (count == 2) {
            res = res + '0';
          } else if (next != 0) {
            count++;
            res = res + next;
          } else {
            res = res + next;
          }
          return res;
        }, '');
        //let pattern = '' + years + days + hours + minutes;
        var _value = format(diffObj1, pattern);
        if (_value && (typeof _value === 'undefined' ? 'undefined' : _typeof(_value)) == 'object') {
          key = _value.key;
          values = getValues(_value.params, diffObj);
          if (pattern.indexOf('0000') == 0) {
            //suffix ignore for second hook
            isSuffixEnable = false;
          } else {
            isSuffixEnable = true;
          }
        } else if (typeof _value == 'string') {
          text = toDateObj.format(_value);
        }
      } else {
        if (isYesterday(toDateObj, fromDateObj)) {
          if (typeof yesterday == 'function') {
            var value = yesterday(diffObj1);
            key = value.key;
            values = getValues(value.params, diffObj);
          } else if ((typeof yesterday === 'undefined' ? 'undefined' : _typeof(yesterday)) == 'object') {
            key = yesterday.key;
            values = getValues(yesterday.params, diffObj);
          } else if (typeof yesterday == 'string') {
            text = toDateObj.format(yesterday);
          }
        } else if (isToday(toDateObj, fromDateObj)) {
          if (typeof today == 'function') {
            var value = today(diffObj1);
            key = value.key;
            values = getValues(value.params, diffObj);
          } else if ((typeof today === 'undefined' ? 'undefined' : _typeof(today)) == 'object') {
            key = today.key;
            values = getValues(today.params, diffObj);
            isSuffixEnable = true;
          } else if (typeof today == 'string') {
            text = toDateObj.format(today);
          }
        } else if (isTomorrow(toDateObj, fromDateObj)) {
          if (typeof tomorrow == 'function') {
            var value = tomorrow(diffObj1);
            key = value.key;
            text = getValues(value.params, diffObj);
          } else if ((typeof tomorrow === 'undefined' ? 'undefined' : _typeof(tomorrow)) == 'object') {
            key = tomorrow.key;
            values = getValues(tomorrow.params, diffObj);
          } else if (typeof tomorrow == 'string') {
            text = toDateObj.format(tomorrow);
          }
        } else {
          var value = others(diffObj1);
          if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) == 'object') {
            key = value.key;
            values = getValues(value.params, diffObj);
            isSuffixEnable = true;
          } else if (typeof value == 'string') {
            text = toDateObj.format(value);
          }
        }
      }
      return text ? React.createElement(
        'span',
        { className: className, 'data-title': title },
        text
      ) : React.createElement(FormatText, {
        i18NKey: isSuffixEnable && suffix != '' ? key + '.' + suffix : key,
        values: values,
        className: className,
        'data-title': title
      });
    }
  }]);

  return DateTimeDiffFormat;
}(React.Component);

export default DateTimeDiffFormat;

DateTimeDiffFormat.propTypes = {
  from: PropTypes.string,
  fromTimeZone: PropTypes.string,
  to: PropTypes.string,
  toTimeZone: PropTypes.string,
  ago: PropTypes.string,
  later: PropTypes.string,
  format: PropTypes.func,
  tommorrow: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.func]),
  yesterday: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.func]),
  today: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.func]),
  others: PropTypes.func,
  title: PropTypes.string
};