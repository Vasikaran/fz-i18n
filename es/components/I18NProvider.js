import _typeof from 'babel-runtime/helpers/typeof';
import _Promise from 'babel-runtime/core-js/promise';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Children } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getI18NValue, userDateFormat } from '../utils';

var emptyObj = {};
var dummy = function dummy(key, values) {
  console.log('Not Mount <I18NProvider/> component.');
  return key;
};
export var i18NProviderUtils = {
  getI18NValue: dummy,
  userDateFormat: dummy
};

var I18NProvider = function (_React$Component) {
  _inherits(I18NProvider, _React$Component);

  function I18NProvider(props, context) {
    _classCallCheck(this, I18NProvider);

    var _this = _possibleConstructorReturn(this, (I18NProvider.__proto__ || _Object$getPrototypeOf(I18NProvider)).call(this, props, context));

    _this.i18n = props.i18n;
    _this.timeZone = props.timeZone;
    _this.user = props.user;
    _this.permission = props.permission;
    _this.license = props.license;
    _this.promise = null;
    _this.resolve = null;
    _this.reject = null;
    _this.state = { reRender: false };
    i18NProviderUtils.getI18NValue = getI18NValue(props.i18n);
    i18NProviderUtils.userDateFormat = userDateFormat(i18NProviderUtils.getI18NValue, props.timeZone);
    return _this;
  }

  _createClass(I18NProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        i18n: this.i18n,
        timeZone: this.timeZone,
        user: this.user,
        permission: this.permission,
        license: this.license
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      if (this.props.i18n != nextProps.i18n || this.props.timeZone != nextProps.timeZone || this.props.user != nextProps.user || this.props.permission != nextProps.permission || this.props.license != nextProps.license && this.promise) {
        this.promise = new _Promise(function (res, rej) {
          _this2.resolve = res;
          _this2.reject = rej;
        }).then(function () {
          _this2.i18n = nextProps.i18n;
          _this2.timeZone = nextProps.timeZone;
          i18NProviderUtils.getI18NValue = getI18NValue(nextProps.i18n);
          i18NProviderUtils.userDateFormat = userDateFormat(i18NProviderUtils.getI18NValue, nextProps.timeZone);
          _this2.setState({ reRender: true }, function () {
            setTimeout(function () {
              return _this2.setState({ reRender: false });
            }, 1);
          });
          _this2.promise = null;
        }, function () {
          _this2.promise = null;
        });
        this.props.onChange(this.resolve, this.reject);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return this.state.reRender ? null : Children.only(this.props.children);
    }
  }]);

  return I18NProvider;
}(React.Component);

export default I18NProvider;


I18NProvider.defaultProps = {
  i18n: emptyObj,
  timeZone: '',
  user: null,
  permission: null,
  license: null
};
I18NProvider.childContextTypes = {
  i18n: PropTypes.object,
  timeZone: PropTypes.string,
  user: PropTypes.object,
  permission: PropTypes.object,
  license: PropTypes.object
};

export var ConnectI18NProvider = connect(function (state, props) {
  var i18n = {};
  var timeZone = '';
  if (typeof props.i18n == 'function') {
    i18n = props.i18n(state);
  } else if (_typeof(props.i18n) == 'object') {
    i18n = props.i18n;
  }
  if (typeof props.timeZone == 'function') {
    timeZone = props.timeZone(state);
  } else if (_typeof(props.timeZone) == 'object') {
    timeZone = props.timeZone;
  }
  return {
    i18n: i18n,
    timeZone: timeZone
  };
})(I18NProvider);

export var AppProvider = connect(function (state, props) {
  var i18n = null;
  var timeZone = '';
  var user = null;
  var permission = null;
  if (typeof props.i18n == 'function') {
    i18n = props.i18n(state);
  } else if (_typeof(props.i18n) == 'object') {
    i18n = props.i18n;
  }
  if (typeof props.timeZone == 'function') {
    timeZone = props.timeZone(state);
  } else if (_typeof(props.timeZone) == 'object') {
    timeZone = props.timeZone;
  }
  if (typeof props.user == 'function') {
    user = props.user(state);
  } else if (_typeof(props.user) == 'object') {
    user = props.user;
  }
  if (typeof props.permission == 'function') {
    permission = props.permission(state);
  } else if (_typeof(props.permission) == 'object') {
    permission = props.permission;
  }
  if (typeof props.license == 'function') {
    license = props.license(state);
  } else if (_typeof(props.license) == 'object') {
    license = props.license;
  }
  return {
    i18n: i18n,
    timeZone: timeZone,
    user: user,
    permission: permission,
    license: license
  };
})(I18NProvider);