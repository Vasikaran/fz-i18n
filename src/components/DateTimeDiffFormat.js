import React, { Children } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import {
  pad,
  getSuffix,
  getValues,
  isToday,
  isYesterday,
  isTomorrow,
  isWithinAWeek
} from '../utils';
import FormatText from './FormatText';

export default class DateTimeDiffFormat extends React.Component {
  constructor(props) {
    super(props);
    this.getSuffix = this.getSuffix.bind(this);
  }
  getSuffix(fromDateObj, toDateObj) {
    let suffix;
    if (toDateObj.isBefore(fromDateObj)) {
      suffix = this.props.ago || '';
    } else if (toDateObj.isAfter(fromDateObj)) {
      suffix = this.props.later || '';
    } else {
      suffix = '';
    }
    return suffix;
  }

  render() {
    const {
      from,
      fromTimeZone,
      to,
      toTimeZone,
      today,
      yesterday,
      tomorrow,
      others,
      format,
      className = null,
      title = null
    } = this.props;
    let fromDateObj = moment(from).tz(fromTimeZone);
    let toDateObj = moment(to).tz(toTimeZone);
    let suffix = this.getSuffix(fromDateObj, toDateObj);
    let diff = moment.duration(Math.abs(toDateObj.diff(fromDateObj)));
    var yearRemaingdays;
    if (toDateObj.isBefore(fromDateObj)) {
      var tempTo = moment(to)
        .tz(toTimeZone)
        .add(diff.get('y'), 'years');
      yearRemaingdays = Math.abs(tempTo.diff(fromDateObj, 'days'));
    } else {
      var tempFrom = moment(fromDateObj)
        .tz(fromTimeZone)
        .add(diff.get('y'), 'years');
      yearRemaingdays = Math.abs(tempFrom.diff(toDateObj, 'days'));
    }
    let withInAWeak = isWithinAWeek(fromDateObj, toDateObj);

    var diffObj = {
      h: diff.get('h'),
      m: diff.get('m'),
      s: diff.get('s'),
      M: diff.get('M'),
      y: diff.get('y'),
      d: diff.get('d'),
      hh: pad(diff.get('h'), 2),
      mm: pad(diff.get('m'), 2),
      ss: pad(diff.get('s'), 2),
      MM: pad(diff.get('M'), 2),
      yy: pad(diff.get('y'), 2),
      dd: pad(diff.get('d'), 2),
      days: yearRemaingdays,
      yDays: pad(yearRemaingdays, 2),
      isWithInAWeek: withInAWeak,
      suffix: suffix
    };
    var diffObj1 = {
      hours: diff.get('h'),
      minutes: diff.get('m'),
      seconds: diff.get('s'),
      months: diff.get('M'),
      years: diff.get('y'),
      days: diff.get('d'),
      yDays: yearRemaingdays,
      isWithInAWeek: withInAWeak,
      suffix: suffix
    };
    //var daysDiff = toDateObj.diff(fromDateObj, 'days', true);
    var key = '';
    var values = [];
    var text = null;
    var isSuffixEnable = false;
    if (format) {
      let years, months, days, hours, minutes, seconds;
      years = diffObj1.years > 1 ? '2' : diffObj1.years;
      //  months = diffObj1.months > 1 ? '2' : diffObj1.months;
      //days = diffObj1.days > 1 ? '2' : diffObj1.days;
      days = yearRemaingdays > 1 ? '2' : yearRemaingdays;
      hours = diffObj1.hours > 1 ? '2' : diffObj1.hours;
      minutes = diffObj1.minutes > 1 ? '2' : diffObj1.minutes;
      //seconds = diffObj1.seconds > 1 ? '2' : diffObj1.seconds;
      // let pattern = '' + years + months + days + hours + minutes + seconds;
      let count = 0;
      let pattern = [years, days, hours, minutes].reduce((res, next) => {
        if (count == 2) {
          res = res + '0';
        } else if (next != 0) {
          count++;
          res = res + next;
        } else {
          res = res + next;
        }
        return res;
      }, '');
      //let pattern = '' + years + days + hours + minutes;
      let value = format(diffObj1, pattern);
      if (value && typeof value == 'object') {
        key = value.key;
        values = getValues(value.params, diffObj);
        if (pattern.indexOf('0000') == 0) {
          //suffix ignore for second hook
          isSuffixEnable = false;
        } else {
          isSuffixEnable = true;
        }
      } else if (typeof value == 'string') {
        text = toDateObj.format(value);
      }
    } else {
      if (isYesterday(toDateObj, fromDateObj)) {
        if (typeof yesterday == 'function') {
          var value = yesterday(diffObj1);
          key = value.key;
          values = getValues(value.params, diffObj);
        } else if (typeof yesterday == 'object') {
          key = yesterday.key;
          values = getValues(yesterday.params, diffObj);
        } else if (typeof yesterday == 'string') {
          text = toDateObj.format(yesterday);
        }
      } else if (isToday(toDateObj, fromDateObj)) {
        if (typeof today == 'function') {
          var value = today(diffObj1);
          key = value.key;
          values = getValues(value.params, diffObj);
        } else if (typeof today == 'object') {
          key = today.key;
          values = getValues(today.params, diffObj);
          isSuffixEnable = true;
        } else if (typeof today == 'string') {
          text = toDateObj.format(today);
        }
      } else if (isTomorrow(toDateObj, fromDateObj)) {
        if (typeof tomorrow == 'function') {
          var value = tomorrow(diffObj1);
          key = value.key;
          text = getValues(value.params, diffObj);
        } else if (typeof tomorrow == 'object') {
          key = tomorrow.key;
          values = getValues(tomorrow.params, diffObj);
        } else if (typeof tomorrow == 'string') {
          text = toDateObj.format(tomorrow);
        }
      } else {
        var value = others(diffObj1);
        if (typeof value == 'object') {
          key = value.key;
          values = getValues(value.params, diffObj);
          isSuffixEnable = true;
        } else if (typeof value == 'string') {
          text = toDateObj.format(value);
        }
      }
    }
    return text ? (
      <span className={className} data-title={title}>
        {text}
      </span>
    ) : (
      <FormatText
        i18NKey={isSuffixEnable && suffix != '' ? key + '.' + suffix : key}
        values={values}
        className={className}
        data-title={title}
      />
    );
  }
}
DateTimeDiffFormat.propTypes = {
  from: PropTypes.string,
  fromTimeZone: PropTypes.string,
  to: PropTypes.string,
  toTimeZone: PropTypes.string,
  ago: PropTypes.string,
  later: PropTypes.string,
  format: PropTypes.func,
  tommorrow: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func
  ]),
  yesterday: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func
  ]),
  today: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func
  ]),
  others: PropTypes.func,
  title: PropTypes.string
};
