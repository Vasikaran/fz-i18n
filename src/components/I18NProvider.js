import React, { Children } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getI18NValue, userDateFormat } from '../utils';

const emptyObj = {};
const dummy = (key, values) => {
  console.log('Not Mount <I18NProvider/> component.');
  return key;
};
export const i18NProviderUtils = {
  getI18NValue: dummy,
  userDateFormat: dummy
};
export default class I18NProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.i18n = props.i18n;
    this.timeZone = props.timeZone;
    this.user = props.user;
    this.permission = props.permission;
    this.license = props.license;
    this.promise = null;
    this.resolve = null;
    this.reject = null;
    this.state = { reRender: false };
    i18NProviderUtils.getI18NValue = getI18NValue(props.i18n);
    i18NProviderUtils.userDateFormat = userDateFormat(
      i18NProviderUtils.getI18NValue,
      props.timeZone
    );
  }
  getChildContext() {
    return {
      i18n: this.i18n,
      timeZone: this.timeZone,
      user: this.user,
      permission: this.permission,
      license: this.license
    };
  }
  componentWillReceiveProps(nextProps) {
    if (
      this.props.i18n != nextProps.i18n ||
      this.props.timeZone != nextProps.timeZone ||
      this.props.user != nextProps.user ||
      this.props.permission != nextProps.permission ||
      (this.props.license != nextProps.license && this.promise)
    ) {
      this.promise = new Promise((res, rej) => {
        this.resolve = res;
        this.reject = rej;
      }).then(
        () => {
          this.i18n = nextProps.i18n;
          this.timeZone = nextProps.timeZone;
          i18NProviderUtils.getI18NValue = getI18NValue(nextProps.i18n);
          i18NProviderUtils.userDateFormat = userDateFormat(
            i18NProviderUtils.getI18NValue,
            nextProps.timeZone
          );
          this.setState({ reRender: true }, () => {
            setTimeout(() => this.setState({ reRender: false }), 1);
          });
          this.promise = null;
        },
        () => {
          this.promise = null;
        }
      );
      this.props.onChange(this.resolve, this.reject);
    }
  }
  render() {
    return this.state.reRender ? null : Children.only(this.props.children);
  }
}

I18NProvider.defaultProps = {
  i18n: emptyObj,
  timeZone: '',
  user: null,
  permission: null,
  license: null
};
I18NProvider.childContextTypes = {
  i18n: PropTypes.object,
  timeZone: PropTypes.string,
  user: PropTypes.object,
  permission: PropTypes.object,
  license: PropTypes.object
};

export const ConnectI18NProvider = connect((state, props) => {
  var i18n = {};
  var timeZone = '';
  if (typeof props.i18n == 'function') {
    i18n = props.i18n(state);
  } else if (typeof props.i18n == 'object') {
    i18n = props.i18n;
  }
  if (typeof props.timeZone == 'function') {
    timeZone = props.timeZone(state);
  } else if (typeof props.timeZone == 'object') {
    timeZone = props.timeZone;
  }
  return {
    i18n,
    timeZone
  };
})(I18NProvider);

export const AppProvider = connect((state, props) => {
  var i18n = null;
  var timeZone = '';
  var user = null;
  var permission = null;
  if (typeof props.i18n == 'function') {
    i18n = props.i18n(state);
  } else if (typeof props.i18n == 'object') {
    i18n = props.i18n;
  }
  if (typeof props.timeZone == 'function') {
    timeZone = props.timeZone(state);
  } else if (typeof props.timeZone == 'object') {
    timeZone = props.timeZone;
  }
  if (typeof props.user == 'function') {
    user = props.user(state);
  } else if (typeof props.user == 'object') {
    user = props.user;
  }
  if (typeof props.permission == 'function') {
    permission = props.permission(state);
  } else if (typeof props.permission == 'object') {
    permission = props.permission;
  }
  if (typeof props.license == 'function') {
    license = props.license(state);
  } else if (typeof props.license == 'object') {
    license = props.license;
  }
  return {
    i18n,
    timeZone,
    user,
    permission,
    license
  };
})(I18NProvider);
