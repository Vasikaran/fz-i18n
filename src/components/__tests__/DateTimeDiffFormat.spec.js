import DateTimeDiffFormat from '../DateTimeDiffFormat';
import I18NProvider from '../I18NProvider';
import React from 'react';
import renderer from 'react-test-renderer';

describe('DateTimeDiffFormat component', () => {
  it("Should display today key", ()=> { 
   var ele=renderer.create(
    <I18NProvider i18n={{today:"today",yesterday:"yesterday"}}>
      <DateTimeDiffFormat from="2016-10-25T05:55:28.000Z" fromTimeZone="Asia/Kolkata" to="2016-10-25T05:55:28.000Z" toTimeZone="Asia/Kolkata" 
        today={{key:"today",params:["hh","mm","ss"]}}
        yesterday={{key:"yesterday",params:["hh","mm","ss"]}}
        tomorrow={{key:"tomorrow",params:["hh","mm","ss"]}}
        others={({years, days, hours, minutes})=>{
          if( days > 7){
            return "DD-MM-YYYY"
          }else{
            return getDateKeyWithParam(years, days, hours, minutes)
          }
        }}
        ago="ago"
        later="later"
        />
    </I18NProvider>
  );
    var tree=ele.toJSON();
    expect(tree).toMatchSnapshot();
  })
  it("Should display today1", ()=> { 
    var toDates=[
      "2016-10-24T05:55:28.000Z",
      "2016-10-24T03:55:28.000Z",
      "2016-10-23T08:55:28.000Z",
      "2016-10-23T08:55:27.000Z",
      "2016-10-25T05:55:28.000Z",
      "2016-10-25T05:55:27.000Z",
      "2016-10-25T05:55:29.000Z",
      "2016-10-25T05:55:28.000Z"
    ]
   var ele=renderer.create(
    <I18NProvider i18n={{today:"today",yesterday:"yesterday",now:"noooow",
    "support.1day.ago":"1day ago","support.1day.nhours.ago":"one day {0} hours ago"}}><div>
      {
      toDates.map((to)=>{
        return <DateTimeDiffFormat from="2016-10-25T05:55:28.000Z" fromTimeZone="Asia/Kolkata" to={to} toTimeZone="Asia/Kolkata" 
        format={({years,days,months,hours},pattern)=>{
          
          switch(pattern)
           {
            case "000000":
              return {
                key:"now"
              }
            break;
            case "001000":
            return {
              key:"support.1day"
            }
            break;
            case "001200":
            case "001201":
            return {
              key:"support.1day.nhours",
              params:["h"]
            }
            break;
          }

        }}
        ago="ago"
        later="later"
        />  
      })
      
      }
      </div>
    </I18NProvider>
  );
    var tree=ele.toJSON();
    expect(tree).toMatchSnapshot();
  })
  it("Should display today later", ()=> { 
   var ele=renderer.create(
    <I18NProvider i18n={{today:"today",yesterday:"yesterday","today.later":"{0}:{1}:{2} later"}}>
      <DateTimeDiffFormat from="2016-10-25T05:55:28.000Z" fromTimeZone="Asia/Kolkata" to="2016-10-25T05:55:29.000Z" toTimeZone="Asia/Kolkata" 
        today={{key:"today",params:["hh","mm","ss"]}}
        yesterday={{key:"yesterday",params:["hh","mm","ss"]}}
        tomorrow={{key:"tomorrow",params:["hh","mm","ss"]}}
        others={({years, days, hours, minutes})=>{
          if( days > 7){
            return "DD-MM-YYYY"
          }else{
            return getDateKeyWithParam(years, days, hours, minutes)
          }
        }}
        ago="ago"
        later="later"
        />
    </I18NProvider>
  );
    var tree=ele.toJSON();
    expect(tree).toMatchSnapshot();
  })
   it("Should display today ago", ()=> { 
   var ele=renderer.create(
    <I18NProvider i18n={{today:"today",yesterday:"yesterday","today.later":"{0}:{1}:{2} later","today.ago":"{0}:{1}:{2} ago"}}>
      <DateTimeDiffFormat from="2016-10-25T05:55:30.000Z" fromTimeZone="Asia/Kolkata" to="2016-10-25T05:55:29.000Z" toTimeZone="Asia/Kolkata" 
        today={{key:"today",params:["hh","mm","ss"]}}
        yesterday={{key:"yesterday",params:["hh","mm","ss"]}}
        tomorrow={{key:"tomorrow",params:["hh","mm","ss"]}}
        others={({years, days, hours, minutes})=>{
          if( days > 7){
            return "DD-MM-YYYY"
          }else{
            return getDateKeyWithParam(years, days, hours, minutes)
          }
        }}
        ago="ago"
        later="later"
        />
    </I18NProvider>
  );
    var tree=ele.toJSON();
    expect(tree).toMatchSnapshot();
  })
 
})