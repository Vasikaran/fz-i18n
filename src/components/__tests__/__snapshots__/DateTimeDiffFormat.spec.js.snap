exports[`DateTimeDiffFormat component Should display today ago 1`] = `
<span>
  00:00:01 ago
</span>
`;

exports[`DateTimeDiffFormat component Should display today key 1`] = `
<span>
  today
</span>
`;

exports[`DateTimeDiffFormat component Should display today later 1`] = `
<span>
  00:00:01 later
</span>
`;

exports[`DateTimeDiffFormat component Should display today1 1`] = `
<div>
  <span>
    1day ago
  </span>
  <span>
    one day 2 hours ago
  </span>
  <span>
    one day 21 hours ago
  </span>
  <span>
    one day 21 hours ago
  </span>
  <span>
    noooow
  </span>
  <span>
    
  </span>
  <span>
    
  </span>
  <span>
    noooow
  </span>
</div>
`;
